src/script/miniscript.h

```diff
@@ -833,51 +858,152 @@ struct Node {
+            case Fragment::WRAP_V: return {{subs[0]->ss.sat.size, subs[0]->ss.sat.diff - 1, subs[0]->ss.sat.exec}, {}};
+            case Fragment::WRAP_J: return {{subs[0]->ss.sat.size, subs[0]->ss.sat.diff, subs[0]->ss.sat.exec | 1}, {1, 0, subs[0]->ss.dsat.exec | 1}};
             case Fragment::THRESH: {
-                auto sats = Vector(internal::MaxInt<uint32_t>(0));
-                for (const auto& sub : subs) {
-                    auto next_sats = Vector(sats[0] + sub->ss.dsat);
-                    for (size_t j = 1; j < sats.size(); ++j) next_sats.push_back((sats[j] + sub->ss.dsat) | (sats[j - 1] + sub->ss.sat));
-                    next_sats.push_back(sats[sats.size() - 1] + sub->ss.sat);
-                    sats = std::move(next_sats);
+                // The maximum satisfaction witness stack size for this fragment, indexed by threshold.
+                auto sats_size = Vector(internal::MaxInt<uint32_t>(0));
+                // Same for the stack size differential after executing this fragment.
+                auto sats_diff = Vector(internal::MaxInt<int32_t>(0));
+                // The maximum stack size at any point during execution of this fragment (or any sub
+                // fragment thereoff), indexed by threshold.
+                auto sats_exec = Vector(internal::MaxInt<int32_t>(0));
+                for (size_t i = 0; i < subs.size(); ++i) {
+                    // Start with the (non-malleable) dissatisfaction size.
+                    auto next_sats_size = Vector(sats_size[0] + subs[i]->ss.dsat.size);
+                    // Whether to account for OP_ADD that drops one element from the stack (-2 + 1).
+                    auto add_diff = (i > 0 && i < subs.size() - 1) ? -1 : 0;
+                    // The total difference in stack size after dissatisfying the first i first fragments.
+                    auto next_sats_diff = Vector(sats_diff[0] + subs[i]->ss.dsat.diff + add_diff);
+                    // Does the stack size gets bigger when dissatifying this fragment than the previous ones?
+                    auto next_sats_exec = Vector(sats_exec[0] | (sats_diff[0] + subs[i]->ss.dsat.exec));
+                    // Update what combination of sat/dsat is the largest per threshold.
+                    for (size_t j = 1; j < sats_size.size(); ++j) {
+                        // Which is larger, k-1 satisfaction and satifying this sub, or k satisfaction and dissatisfying
+                        // this one?
+                        next_sats_size.push_back((sats_size[j] + subs[i]->ss.dsat.size) | (sats_size[j - 1] + subs[i]->ss.sat.size));
+                        // Same for size of the stack after satisfying k of the past i fragments.
+                        next_sats_diff.push_back((sats_diff[j] + subs[i]->ss.dsat.diff) | (sats_diff[j - 1] + subs[i]->ss.sat.diff) + add_diff);
```

Comment:
"warning: suggest parentheses around arithmetic in operand of ‘|’ [-Wparentheses]"

```suggestion
                        next_sats_diff.push_back((sats_diff[j] + subs[i]->ss.dsat.diff) | ((sats_diff[j - 1] + subs[i]->ss.sat.diff) + add_diff));

Reply:
Thanks, good suggestion that i add those for clarity because your suggestion is incorrect: `add_diff` should always be added. The current (and intended) semantic is:
```diff
diff --git a/src/script/miniscript.h b/src/script/miniscript.h
index 011c7ac6d6..35797153da 100644
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -1002,7 +1002,7 @@ private:
                         // this one?
                         next_sats_size.push_back((sats_size[j] + subs[i]->ss.dsat.size) | (sats_size[j - 1] + subs[i]->ss.sat.size));
                         // Same for size of the stack after satisfying k of the past i fragments.
-                        next_sats_diff.push_back((sats_diff[j] + sub_diff.dsat) | (sats_diff[j - 1] + sub_diff.sat) + add_diff);
+                        next_sats_diff.push_back(((sats_diff[j] + sub_diff.dsat) | (sats_diff[j - 1] + sub_diff.sat)) + add_diff);
                         // Update the maximum stack size for satisfying k subs with the largest of the maximum cost of
                         // satisfying k-1 subs (implying we satisfy this one), the cost of dissatisfying this sub
                         // after the satisfaction of j subs beforehand, and the cost of satisfying this sub after the
```

Did you get this warning with GCC?